### PR TITLE
Make node-libs-browser optional

### PR DESCRIPTION
**Pull request by baderbuddy**
**What kind of change does this PR introduce?**

Feature

**Did you add tests for your changes?**

No new tests

**If relevant, link to documentation update:**

**Summary**
Make node-libs-browser optional.
#3058

**Does this PR introduce a breaking change?**

Yes, since node-libs-browser is now optional, anyone that was depending on the NodeSourcePlugin and also has optional dependencies turned off will no longer work, the user will have to install this optional dependency.

**Other information**
I've run the tests locally now and everything appears to run okay.

### CODE REVIEW

**Comment by TheLarkInn** [comment]
Any way you could add a test for this?

**Comment by baderbuddy** [comment]
I've tried to add some tests as best as I could. It does seem like a hard thing to test :)

File: lib/node/NodeSourcePlugin.js [hunk 1]
```diff
@@ -35,54 +33,60 @@ module.exports = class NodeSourcePlugin {
 				return ParserHelpers.addParsedVariableToModule(this, name, mockModule + suffix);
 			});
 		}
+		try {
```
Comment (sokra) [inline lib/node/NodeSourcePlugin.js]: Don't wrap this is a try-catch block.

File: lib/node/NodeSourcePlugin.js [hunk 2]
```diff
@@ -5,16 +5,14 @@
...
 module.exports = class NodeSourcePlugin {
 	constructor(options) {
 		this.options = options;
 	}
+
 	apply(compiler) {
-		const options = this.options;
-		if(options === false) // allow single kill switch to turn off this plugin
-			return;
+		var nodeLibsBrowser = null;
```
Comment (sokra) [inline lib/node/NodeSourcePlugin.js]: const

File: lib/node/NodeSourcePlugin.js (hunk 1 above)
Comment (baderbuddy) [reply lib/node/NodeSourcePlugin.js]: I'd need the try catch for the require, but I can see it making sense to limit it as much as possible. If we really want to leave nodeLibsBrowser as a constant then the only option I can think of would be to wrap the require in a function that has the try/catch and returns null if it gets an exception. What do you think of that?

**Comment by baderbuddy** [comment]
@sokra I haven't heard back on this in a while, are there any more changes needed?

**Comment by sokra** [comment]
I don't see the benefits of this PR...

**Comment by baderbuddy** [comment]
This library contains some dependencies that aren't allowed where I work. Making it optional means that I can install webpack without node-libs-browser

**Comment by sokra** [comment]
> This library contains some dependencies that aren't allowed where I work. Making it optional means that I can install webpack without node-libs-browser

If this is the only reason, I'm going to close this, as this has no benefit for the general public and increase complexity. You can use yarn resolutions to override node-libs-browser.

**Comment by baderbuddy** [comment]
@sokra yarn resolutions will still cause webpack to fail if node-libs-browser isn't present. This was an accepted issue https://github.com/webpack/webpack/issues/3058 and will make the overall webpack bundle lighter for those that don't install optional dependencies by making a feature that should have been optional in the first place, optional.

**Comment by flisky** [comment]
We're working on a nodejs application. And we have to `"postinstall": "rm -rf node_modules/assert node_modules/util node_modules/querystring"`, because the core library shims brought from `node-libs-browser` are almost outdated and buggy.

**Pull request closed without merging**